LLK/scratch-desktop
# macOS: request camera & microphone permissions on demand

### Resolves

- Resolves #109 

### Proposed Changes

This batch of changes uses `askForMediaAccess` to request camera and microphone access when necessary.

Supporting changes include:

- Add a section in `README.md` mentioning a known code-signing issue which can sometimes interfere with debugging
- Use separate entitlements lists for Mac App Store vs. non-store ("Developer ID") macOS builds
  - Both macOS builds now use the `cs.allow-jit` entitlement since documentation suggests that it's necessary for JavaScript to work well.
  - non-Store mac builds no longer use the `app-sandbox` entitlement since it seems to cause code-signing problems. Mac App Store builds still use it as it's required for the store.
- Add descriptions for why the app asks for camera and microphone permissions, the consequences of denying those permissions, and how to change your mind.

### Reason for Changes

On macOS 10.14 (Mojave) and macOS 10.15 (Catalina), apps must request access to the microphone or camera. Previous versions of Scratch Desktop didn't request permission, so camera and microphone features didn't work correctly.

### Test Coverage

Manually tested. Example test sequences below.

Note that the `tccutil reset Camera` and/or "tccutil reset Microphone` command(s) must be used before each test if you would like to see the prompt again. You can also enable or disable permission through System Preferences.

Camera:
- Run `tccutil reset Camera` (this will reset camera permissions for ALL applications)
- Start this build of Scratch Desktop
- Attempt to add a costume by taking a picture, or add the "video sensing" extension
- You should get a pop-up dialog asking for permission
   - If you deny the access the "video sensing" extension and sound recorder won't work. You should receive a dialog explaining why the application needs these features.
   - If you allow the access the video sensing extension and taking a photo in the paint editor should work correctly.

Microphone:
- Run `tccutil reset Microphone` (this will reset microphone permissions for ALL applications)
- Start this build of Scratch Desktop
- Attempt to record a sound in the sound editor or add the "loudness" block to the stage
- You should get a pop-up dialog asking for permission
   - If you deny the access the loudness block and sound recorder won't work. You should receive a dialog explaining why the application needs these features.
   - If you allow the access the loudness block and sound recorder should work correctly.
